# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x + 2
print(addTwo(4))

# Lambda Function Syntax
# Example of calling a Lambda function without a variable
print((lambda x: x + 2)(4))  #Good for a quick output.  
                              #its anonymous function not assigned to a variable yet?


6
6


#### Saving to a Variable

In [3]:
f_test = lambda x:x + 2  # this way takes up less memory.This is how you can use it as a variable


f_test(4)

6

#### Multiple Inputs

In [5]:
# Multiple Inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

# Multiple Inputs with a variable attached
x_text = lambda x,y,z: x*y*z
print(x_text(3,5,8))

120
120


#### Passing a Lambda into a Function

In [6]:
def multiply(f,num): # f is =lambda function
                    # num is = 4
    """
        f expects a lambda function
        num expects a number    
    """
    return f(num) 
multiply(lambda x: x * x, 4) # Lambda 4: 4 * 4

16

#### Returning a Lambda from a Function

In [22]:
# Regular defined function
def multiply_test(num):
    return num * 4

# Function within a function
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply #add"multiply,test" to return multiple variables yields a tuple

f_return = returnFunc()  # This step allows us to have access to both levels by combining them
print(f_return(4)) # to return the function with in a function


#Lambda function returned from a regular function (using 4 parameters instead of just 2)
def returnLam(b,c):
    return lambda x,a: x + a + b + c
r_lamb = returnLam(4,5) # b=4 and c=5
print(r_lamb(5,5)) #x=5 and a=5


8
19


#### If Statements within Lambdas

In [25]:
# Lambda x : True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2  
                                                # Do the multiplcation if the number is > 10, 
                                                # otherwise/else add it by 2
print(f_condition(8))

print(f_condition(12))

print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [28]:
f = lambda cube: cube ** 3  #annonymous function it's always going to be a lambda function

print(f(2))
print(f(3))
print(f(4))

8
27
64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [29]:
# map(func,iterable(list,dict,tuple,set))
# Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 **2
    else:
        return num, num2
    
numbers = [4,11,20,3,15]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)


[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [31]:
# map(lambda x: x + 2, list)
# using lambda in map happens in line (or in one line) usually
# list(map(lambda x : x * 2,nums))
squared_nums_lamb=list(map(lambda x,y: (x**2, y**2) if x<10 and y<10 else (x,y), numbers, more_nums))

print(squared_nums_lamb)



[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [33]:
double_num = list(map(lambda x,y: (((x*2)-1),((y*2)-1)) if x>0 and y>0 else (x,y), numbers, more_nums))
print(double_num)

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [34]:
# Boolean mask - if i had to rewrite this information, the function would be looking for a true or
# false result.  otherwise it will be dropped.
 #          F     T      F     F       T
names = ['Bob','Andy','Max','Evan','Angelica']

def a_names(name):
    if name[0].lower() == 'a': # only checking first index/letter
        return True
    else:
        return False
    
new_names_list = list(filter(a_names,names)) # runs list of names in function a_names yields only true values
print(new_names_list)

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [36]:
new_names_lamb=list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [47]:
from statistics import mean

nums = [2,7,4.2,1.6,9,4.4,4.9]
print(mean(nums))
mean1=(mean(nums))

# Lambda function Solution
m = list(filter(lambda num: True if num<mean1 else False, nums))
print(m)

# regular function Solution
def numbers(num):
        if num<mean1:
            return True
        else:
            False

below_numbers = list(filter(numbers,nums)) 
print(below_numbers)        

4.728571428571429
[2, 4.2, 1.6, 4.4]
[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [51]:
from functools import reduce  #starts from left and takes first pairs and does what you want. 
                            #takes that total and does that to the next value..process continues

# reduce(function,iterable)

list_1 = [2,4,6,8,10]

def addNums(num1, num2):  #num1, after 1st iteration, will continue to be the running total
    return num1 + num2

result_add = reduce(addNums, list_1)

print(result_add)

# Subtract a list of numbers
def subtractNums(num1,num2): #num1, after 1st iteration, will continue to be the running total
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)


30
-26


#### Using Lambda's with Reduce()

In [52]:
result_lamb = reduce(lambda x,y: x+y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [54]:
my_list = [1,2,3,4]

from functools import reduce

result_lamb = reduce(lambda x,y: x*y, my_list)
print(result_lamb)



24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [57]:
def addNums(num):
    # Set base case for recursive function
    if num <= 1:
        print("addNums(1) = 1")
        return num #when the function reaches this point, it'll THEN ADD THE VALUES UPWARD (1+2+3+4+5) = 15
    else:
        print(f"addNums({num}) = {num} + addNums({num-1})")
        return num + addNums(num - 1)  #decreasing the value of num here and starting at
                                    # addNums again with num-1 value
#after the loop of decreasing by 1 is done, the num's are then added by the call stack at line 5
addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [63]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <=1:
        return 1 
    else:
        return num * factorial(num-1) # loops it back to start here
            # '*' tells it to keep multiplying for the final oupout, happens at line 4
    
factorial(5)


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [66]:
# return num - subNums(num - 1)

def subtract(x):
    if x <=1:
        return x
    else:
        return x - subtract(x-1)

subtract(10)
    

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [67]:
def my_range(stop, start, step = 1):
    while start < stop:
        yield start #Yield keyword denotes a generator.  This is where it all happens!
        start += step
        
for i in my_range(20, start= 2): #Yield needs a 'for' loop to work
    print(i)
    


2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [1]:
# for i in numbers_squared(10)

def numbers_squared(stop, start, step = 1):
    while start > stop:
        yield start**2
        start -= step #step function       
        
for i in numbers_squared(0, start=3): #Yield needs a 'for' loop to work
    print(i)

9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [15]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]


def spaces(name):
        if name != '' and name !=' ' and name !='  ':
            return True
        else: 
            return False
new_places = list(filter(spaces, places))
print(new_places)
        


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [27]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
# lambda function, use .sort() to access all values
# after sorting, the output should be as above.
# use split and sort by last name (second index)


author.sort(key=lambda name: name.split(" ")[-1].lower())
print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [26]:
# F = (9/5)*C + 32, lambda function, keep these as tuples to manipulate
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

converted=map(lambda x[1]: (9/5)*x[1] +32 , places)  #trying to access the second position of each tuple. smh

print(converted)



TypeError: 'list' object is not callable

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [17]:
# output - positionnumber: fibonacci total

# count=0
# n1=0 
# n2=1
# def fibonacci(num):
#         if num<=0:
#             return num
#         else:
#             while count < num:
#                 print(f'Iteration {num}: {n1}')
#                 nth = n1 + n2
#                    # update values
#                 n1 = n2
#                 n2 = nth
#                 count += 1
#             return num + fibonacci(num-1)
        
# fibonacci(5)

def fibonacci(num):
        if num <= 1:
            return num
        else:
            print(f'Iteration {num}: {num-1}')
            return(fibonacci(num-1) + fibonacci(num-2))

fibonacci(5)
   

Iteration 5: 4
Iteration 4: 3
Iteration 3: 2
Iteration 2: 1
Iteration 2: 1
Iteration 3: 2
Iteration 2: 1


5